## Chapter 2 ##
## Basic RNA-Seq concepts ##

**2.1 What is RNA-Seq when simplified to its essence?**

Conceptually an RNA-Seq analysis is quite straightforward process. It goes like this:

 1. We produce sequencing data from a transcriptome in a “control” state
 2. We match sequencing reads to the genome or the transcriptome.
 3. We count how many reads align to a region (feature). Let’s say 100 reads overlap with Gene A.
 
Now say the cell is subjected to a different condition, for example, a cold
shock is applied. We perform the same measurements as before:

 1. We produce sequencing data from the transcriptome in the “perturbed” state
 2. We match sequencing reads to the genome or the transcriptome.
 3. We count how many reads align to the same region (feature), gene A.

Suppose 200 reads overlap with Gene A.We see that 200 is twice as big as 100.

file:///home/preonath/Pictures/Screenshot%20from%202021-03-03%2013-34-10.png




Since the coverage is proportional to the abundance of the transcript (expression level), we can report that the
transcription level for gene A doubled in the second experiment. We call this doubling the differential expression.

We need to investigate Gene A, what functions of it are known, is the response to cold shock one of them? If yes -what is the story, if not perhaps we discovered a new role for the gene. And so on.


**2.2 What makes RNA-Seq challenging?**

The stochasticity, and imprecision of the many processes coupled to the scale of data conspire against determining fold change so readily. We can’t just divide the two numbers; we can’t even trust the numbers to mean what they are supposed to. There is going to be a lot of bookkeeping, defensive measures taken to protect against you-know-who: The Dark Lord of False Discoveries. He grows stronger as the data grows; thus the defense is challenging even though that task itself is quite straightforward: compare one number to another and estimate whether their difference is big enough to be considered a valid change.

Take solace and strength from recognizing that the complexity lies in the data and information management - the analysis process itself does not call upon particularly abstract, mathematical or analytical problem-solving skills.On the other side, you best understand early on that there is no silver bullet that would make everything always fall into place. There is no magic, always correct approach that will always guarantee a correct data analysis. Your job will always be to respond to what the data tells you. For that, you will need to understand the methods in a little more depth

**How does RNA sequencing work?**

- The RNA-seq protocol turns the RNA produced by a cell into DNA (cDNA,complementary DNA) via a process known as reverse transcription.

- Theresulting DNA is then sequenced, and from the observed abundances of DNA,we attempt to infer the original amounts of RNA in the cell.

- It seems that all we need to do is countis how many DNA fragments are there. If we get higher numbers under aparticular condition, it means that the cell must be producing RNA at highelevels. 

- In reality, the problems of counting and comparing are more compli-cated and confounded by several factors
  - the most important being that themRNA levels in the cell are almost never static. There is a continuous ebband flow, each in response to particular stimuli. 
  
- We need to isolate the signal that corresponds to the function of interest from all the other unrelatedchanges that might be present in the mRNA concentrations.


**What is the final result of an RNA-Seq analysis?**

The result of an RNA-Seq analysis is a quantification matrix. For our toyexample the file might look like this:
        
        name     control    cold-shock
        Gene A   100        200
        Gene B   300        120
        Gene C   400        530
This quantification file when processed by a statistical method will be augmented with statistical measures that characterize the changes over the comparisons that we might choose to make. 
 - For example, if we were to perform a pairwise comparison between cold-shock and control, the file above will gainmore information, expressed as additional columns:

        name     control    cold-shock     fold_change    pvalue
        Gene A   100        200              2.0           0.000035
        Gene B   80         60               0.75          0.234
        Gene C   120        180              1.5           0.013
        
     
        Interpreting this file at the 0.05 statistical significance level would then tell us that Gene A and Gene C are differentially expressed between control and cold-shock, whereas Gene B is not.

## Chapter 3 ##

## 1. Introducing the Golden Snidget ##

- The data that we will analyze was created as so-called spike-in control, where known abundances of mRNA were added to mRNA extracted from a cell; then both the artificial spike-in and the real data were sequenced on an Illumina sequencer.

- The study was originally published as Informatics for RNA-seq: A web resource for analysis on the cloud 1 in PLoS Computational Biology (2015) by Malachi Griffith et al.

- While working through the example data I found the reanalysis and especially the validation of the control data to be unexpectedly tedious.

- In a moment of inspiration, I came up with the idea of transforming the data to represent a more interesting and more realistic scenario. Basically, I have “created” a hypothetical organism that corresponds to the control data. 

- **This organism, the Golden Snidget, that magical golden bird with fully rotational wings, will help you understand how RNA-Seq data analysis works, from its strengths to its limitations**

- The Golden Snidget is not a toy model! You will see that the same code and approach used to analyze the Golden Snidget data can be used almost unchanged in many other realistic data analysis scenarios.

**3.1 What do we know about the Golden Snid-get?**


Partly mechanical and partly fictitious, the Golden Snidget turns out to be the perfect candidate for RNA-Seq data analysis. What makes it so well suited? Two peculiar features. First, the Golden Snidget can only be in two moods:

1. BORED when it flies nice and straight.
2. EXCITED when it constantly changes direction.

Additionally, the Golden Snidget is so magically regulated that, in each state, we know precisely how many transcripts are expressed. The organism is so perfectly controlled that even genes could be named in a way that
describes their changes for both states. For example, a gene might be called:

**ABC-1000-UP-4**

- The gene was named as such because in the BORED state, each cell has 1000 copies of the transcript corresponding to gene ABC. Moreover, the name also conveys that in the EXCITED state, this same gene is be up-regulated 4 fold compared to the BORED state.

- Another way to explain the naming scheme is that in the BORED state, there will be 1000 copies of the ABC transcript, whereas in the EXCITED state, each cell will contain 4000 copies of the ABC transcript.


- Thus, just by looking at a gene name, we already know how much of it is inside the cells. If we were to build a quantification matrix for the real mRNA abundance inside the cell, our data would look like this:


          name            BORED        EXCITED        foldChange
          
       ABC-1000-UP-4      1000         4000            4
       ABD-40-DOWN-2      40           20              2
       ABE-90-SAME-1      90           90              1


- Across the two states each gene can only be either UP, DOWN regulated or stay the SAME. Note how the quantification matrix above represents what “really” happens in the cell.

- It remains to be seen how well results from an RNA-Seq data can reproduce the expected numbers or ratios. The gene naming convention will allow you to spot check any intermediate result and evaluate how well the analysis is going.




**3.2 What is the objective of the entire section on the Golden Snidget?**

- Your job will be to demonstrate how well, and under what circumstances, can you reproduce the known regulatory behavior when starting with RNA-Seq data.

- As you study the Golden Snidget and unlock its secrets, you will begin to understand the bigger picture of how to think about RNA-Seq in general.

- Your first task is to understand the genome that you wish to quantify.

## Chapter 4 ##

## Understand your reference ##



- The very first task of RNA-Seq data analysis is to select the “reference” that your study will use to quantify the gene expression. As we described in the previous chapter, there are two analysis paradigms:

1. Quantify against a genome
2. Classify against a transcriptome

- As you recall, we also recommended that you do both.
   - Your first step will be to identify and obtain the reference files for your study.
   - Typically there are data distribution sites distributing reference files. For the


- Golden Snidget, as far as we know, there is only one source of information, the reference is available from:
    • http://data.biostarhandbook.com/books/rnaseq/data/golden.genome.tar.gz
    
    
**4.1 What is my very first step?**


- Understand your reference data! What properties does the “reference” have?

    **1.How big is the reference?
    2.How many chromosomes does it have?
    3.What are these chromosomes called? How long is each?
    4.How many features are annotated?
    5.What types of features are listed?**



- The process of understanding your genome is the first step in the confidence building process that moves you from being a mere passenger, tagging along to becoming the driver and director of the process.


**4.2 What is the Golden Snidget’s genome like?**


   - First, do yourself a favor, create a subdirectory, and work in that. We have seen countless well-intentioned individuals struggle to make sense of their methods because their home directories were overrun with hundreds of files of various origins. Make a new separate directory for every new project!

- As a matter of fact we recommend creating a main working directory call that work,then creating a new subdirectory in that for every subproject for examplegolden‘, like so:


In [18]:
%%bash
source /home/preonath/anaconda3/etc/profile.d/conda.sh %conda activate base

In [19]:
%%bash

# Make two directories: work/golden
mkdir -p work/golden


In [20]:
%%bash

# Switch to the directory
cd ./work/golden

## We assume that all commands forward on are run in the work/golden directory that you have just created and that the bioinfo environment is activated. Now let’s get the reference data and investigate it

# Download the reference genome.
wget http://data.biostarhandbook.com/books/rnaseq/data/golden.genome.tar.gz
pwd

/home/preonath/Biostar/RNA_Seq/work/golden


--2021-04-01 23:58:09--  http://data.biostarhandbook.com/books/rnaseq/data/golden.genome.tar.gz
Resolving data.biostarhandbook.com (data.biostarhandbook.com)... 198.74.58.207
Connecting to data.biostarhandbook.com (data.biostarhandbook.com)|198.74.58.207|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 57462 (56K) [application/octet-stream]
Saving to: 'golden.genome.tar.gz.1'

     0K .......... .......... .......... .......... .......... 89% 50.0K 0s
    50K ......                                                100% 40.9M=1.0s

2021-04-01 23:58:10 (56.1 KB/s) - 'golden.genome.tar.gz.1' saved [57462/57462]



In [21]:
%%bash

cd ./work/golden

# Unpack the reference genome.
tar xzvf golden.genome.tar.gz

refs/genome.fa
refs/features.gff
refs/transcripts.fa


In [22]:
%%bash
cd ./work/golden

#At this point, you will see that a refs folder has been created from the downloaded data. List the contents of it:

ls -l refs/

#We can see that we have a genome, features, and transcripts. For otherorganisms, the naming of files might not be as obvious and clear cut.


total 6188
-rw-r--r-- 1 preonath preonath   12527 Feb  5  2020 features.gff
-rw-r--r-- 1 preonath preonath  130411 Feb  5  2020 genome.fa
-rw-rw-r-- 1 preonath preonath 4237485 Mar 28 16:36 genome.fa.1.ht2
-rw-rw-r-- 1 preonath preonath   32196 Mar 28 16:36 genome.fa.2.ht2
-rw-rw-r-- 1 preonath preonath      26 Mar 28 16:36 genome.fa.3.ht2
-rw-rw-r-- 1 preonath preonath   32189 Mar 28 16:36 genome.fa.4.ht2
-rw-rw-r-- 1 preonath preonath   62331 Mar 28 16:36 genome.fa.5.ht2
-rw-rw-r-- 1 preonath preonath   32708 Mar 28 16:36 genome.fa.6.ht2
-rw-rw-r-- 1 preonath preonath      12 Mar 28 16:36 genome.fa.7.ht2
-rw-rw-r-- 1 preonath preonath       8 Mar 28 16:36 genome.fa.8.ht2
-rw-rw-r-- 1 preonath preonath      23 Mar 28 16:38 genome.fa.fai
-rw-rw-r-- 1 preonath preonath       0 Mar 28 17:26 temp-core-140304-2C6E85A6505E.sam
-rw-rw-r-- 1 preonath preonath 1644570 Apr  1 22:00 transcript.idx
-rw-rw-r-- 1 preonath preonath   29958 Mar 28 15:28 transcripts.bam
-rw-rw-r-- 1 preonath preonath 

**4.3 How many sequences in the genome?**


In [23]:
%%bash

cd ./work/golden
#Evaluate the FASTA files:

seqkit stats ./refs/*.fa

# The genome has a single chromosome of size of 128,765 bp. There are 92 transcripts. The shortest transcript is 273bp. The longest is 2022bp. 
# The total transcript size is 92 x 899 = 82,708 basepairs 

file                   format  type  num_seqs  sum_len  min_len  avg_len  max_len
./refs/genome.fa       FASTA   DNA          1  128,756  128,756  128,756  128,756
./refs/transcripts.fa  FASTA   DNA         92   82,756      273    899.5    2,022


**4.4 What does the genome file look like?**

In [24]:
%%bash
cd ./work/golden

cat ./refs/genome.fa | head -5

>Golden full genome, version 2020, Yo Ho Ho!
GCATTTTGAAAATTCTATGGAAGAGCTAGCATCTCTGACGAAAACAGCAGACGGAAAAGTACTGACCAGCGTCACACAAA
AACGGAACAGGGCTGACGCCGCTACATATATAGGAAAAGGGAAGGTAGAAGAGCTGAAGGCACTCGTGGAAGAGCTTGAA
GCTGATCTCCTCATCTTTAATGATGAACTGTCGCCAAGTCAGCTGAAGTCATTGGCAACAGCAATTGAAGTGAAGATGAT
TGACCGCACGCAATTGATATTAGATATTTTTGCAAAGCGGGCGAGAACGAGAGAAGGCAAACTTCAAATTGAGCTGGCTC


**4.5 What is the annotation file?**

In [25]:
%%bash
cd ./work/golden
cat ./refs/features.gff | head -5

##gff-version 3
Golden	ERCC	exon	1	1060	0	+	.	gene_name=AAA-750000-UP-4; gene_id=AAA-750000-UP-4; transcript_id=AAA-750000-UP-4-T; exon_number=1;
Golden	ERCC	exon	1560	2083	0	+	.	gene_name=ABA-187500-UP-4; gene_id=ABA-187500-UP-4; transcript_id=ABA-187500-UP-4-T; exon_number=1;
Golden	ERCC	exon	2583	3616	0	+	.	gene_name=ACA-46875-UP-4; gene_id=ACA-46875-UP-4; transcript_id=ACA-46875-UP-4-T; exon_number=1;
Golden	ERCC	exon	4116	5138	0	+	.	gene_name=ADA-23438-UP-4; gene_id=ADA-23438-UP-4; transcript_id=ADA-23438-UP-4-T; exon_number=1;


In [26]:
%%bash
cd ./work/golden
cat ./refs/features.gff | grep exon | wc -l

92


**We note that the chromosomal names in the feature file have the same value:**

- Golden as in the genome. So the two names will match. Mismatching chromosome names, for example if it were called golden instead, can be very annoying to deal with.

**4.6 What do the transcripts look like?**

In [27]:
%%bash
cd ./work/golden
cat ./refs/transcripts.fa | head -5

>AAA-750000-UP-4
GCATTTTGAAAATTCTATGGAAGAGCTAGCATCTCTGACGAAAACAGCAGACGGAAAAGTACTGACCAGCGTCACACAAA
AACGGAACAGGGCTGACGCCGCTACATATATAGGAAAAGGGAAGGTAGAAGAGCTGAAGGCACTCGTGGAAGAGCTTGAA
GCTGATCTCCTCATCTTTAATGATGAACTGTCGCCAAGTCAGCTGAAGTCATTGGCAACAGCAATTGAAGTGAAGATGAT
TGACCGCACGCAATTGATATTAGATATTTTTGCAAAGCGGGCGAGAACGAGAGAAGGCAAACTTCAAATTGAGCTGGCTC


In [29]:
%%bash

cd ./work/golden
#We can see the naming schemes. Print out more gene names:
cat ./refs/transcripts.fa | grep ">" | wc -l

92


- From here, we can see that in this genome gene, AAA expresses with 750,000 copies in the BORED state, whereas gene AJA expresses with 732 copies. 

- Basically, at any time, there will 1000x more transcripts for the AAA than for AJA. Make a mental note of that.

- Do also note how both the AAA and the AJA genes change the same amount four fold! The magnitude of the change (the ratio) across the states for AAA and AJA be the same; yet the absolute amount of the concentration of each transcript will, however, be massively different.

- It makes sense that it will be more challenging to quantify AJA than AAA accurately. There may be a limit under which we cannot detect changes; perhaps we can’t discover anything at levels of AFA and below.

- Of course for any other organism, we would not know beforehand which transcript might be lowly expressed. But we need to approach the analysis with the expectation that we will only be able to detect transcripts above a certain expression abundance threshold.

**4.8 Optional step: align your transcripts against the reference**

You may want to validate your transcripts. Will they match the genome as you assume they would?

In [30]:
%%bash

cd ./work/golden/refs
pwd
# The index to the genome
IDX=genome.fa
hisat2-build $IDX $IDX

# Create the transcript alignment BAM file.
hisat2 -x $IDX -f -U transcripts.fa | samtools sort > transcripts.bam



/home/preonath/Biostar/RNA_Seq/work/golden/refs
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:00:00
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:00
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:00
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Split 1, merged 6; iterating...
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 18392.7 (target: 24140)
Getting block 1 of 7
  Reserving size (24141) for bucket 1
  Calculating Z arrays for bucket 1
  Ente

Settings:
  Output files: "genome.fa.*.ht2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Local offset rate: 3 (one in 8)
  Local fTable chars: 6
  Local sequence length: 57344
  Local sequence overlap between two consecutive indexes: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  genome.fa
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
  Time to read SNPs and splice sites: 00:00:00
Using parameters --bmax 24141 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 24141 --dcv 1024
Constructing suffix-array element generator
Conv

In [31]:
%%bash

cd ./work/golden/refs
# Index the BAM file
samtools index transcripts.bam


**5 3. Understand the sequencing reads**


The next step is to get a solid grasp of the properties and layout of your sequencing reads. If the data is stored in SRA, you can use fastq-dump; for other cases, your data will be stored in files distributed from various locations. In our case, the data is located at:

 - http://data.biostarhandbook.com/books/rnaseq/data/golden.reads.tar.gz


**5.1 What information do I need to know?**


You need to know how many reads were measured, what the experimental layout is, and get a sense of how well the experiment worked. You may have access to various sources of information that describe the experiment.

In general, we found that often documentation and even published supplementary information often lacks essential details and describes data incompletely. As a rule, you can often only continue on by figuring out by yourself, from the files themselves of what they contain. Thus we will follow that strategy:



In [32]:
%%bash

cd ./work/golden

# Download the data
wget http://data.biostarhandbook.com/books/rnaseq/data/golden.reads.tar.gz

# Unpack the data
tar zxvf golden.reads.tar.gz

# List the content of the current directory.
ls -l

# List the contents of the reads directory
ls -l reads


reads/BORED_1_R1.fq
reads/BORED_1_R2.fq
reads/BORED_2_R1.fq
reads/BORED_2_R2.fq
reads/BORED_3_R1.fq
reads/BORED_3_R2.fq
reads/EXCITED_1_R1.fq
reads/EXCITED_1_R2.fq
reads/EXCITED_2_R1.fq
reads/EXCITED_2_R2.fq
reads/EXCITED_3_R1.fq
reads/EXCITED_3_R2.fq
total 234808
drwxrwxr-x 2 preonath preonath      4096 Mar 28 17:05 bam
-rw-rw-r-- 1 preonath preonath      4912 Mar 29 02:20 counts.txt
-rw-rw-r-- 1 preonath preonath       610 Mar 28 23:13 counts.txt.summary
-rw-rw-r-- 1 preonath preonath     57462 Feb  5  2020 golden.genome.tar.gz
-rw-rw-r-- 1 preonath preonath     57462 Feb  5  2020 golden.genome.tar.gz.1
-rw-rw-r-- 1 preonath preonath 120138017 Feb  5  2020 golden.reads.tar.gz
-rw-rw-r-- 1 preonath preonath 120138017 Feb  5  2020 golden.reads.tar.gz.1
-rw-rw-r-- 1 preonath preonath        54 Mar 28 16:39 ids
drwxrwxr-x 8 preonath preonath      4096 Mar 29 02:12 output
drwxrwxr-x 2 preonath preonath      4096 Apr  2 00:24 reads
drwxrwxr-x 2 preonath preonath      4096 Apr  1 23:58 refs

--2021-04-01 23:59:31--  http://data.biostarhandbook.com/books/rnaseq/data/golden.reads.tar.gz
Resolving data.biostarhandbook.com (data.biostarhandbook.com)... 198.74.58.207
Connecting to data.biostarhandbook.com (data.biostarhandbook.com)|198.74.58.207|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 120138017 (115M) [application/octet-stream]
Saving to: 'golden.reads.tar.gz.1'

     0K .......... .......... .......... .......... ..........  0% 79.4K 24m36s
    50K .......... .......... .......... .......... ..........  0%  171K 18m1s
   100K .......... .......... .......... .......... ..........  0% 78.6K 20m18s
   150K .......... .......... .......... .......... ..........  0%  163K 18m12s
   200K .......... .......... .......... .......... ..........  0%  175K 16m47s
   250K .......... .......... .......... .......... ..........  0% 92.8K 17m29s
   300K .......... .......... .......... .......... ..........  0%  169K 16m38s
   350K .......... .......... ....

In [33]:
%%bash

cd ./work/golden
#Run statistics on the reads:

seqkit stats ./reads/*.fq

file                     format  type  num_seqs     sum_len  min_len  avg_len  max_len
./reads/BORED_1_R1.fq    FASTQ   DNA    112,193  11,219,300      100      100      100
./reads/BORED_1_R2.fq    FASTQ   DNA    112,193  11,219,300      100      100      100
./reads/BORED_2_R1.fq    FASTQ   DNA    137,581  13,758,100      100      100      100
./reads/BORED_2_R2.fq    FASTQ   DNA    137,581  13,758,100      100      100      100
./reads/BORED_3_R1.fq    FASTQ   DNA    123,093  12,309,300      100      100      100
./reads/BORED_3_R2.fq    FASTQ   DNA    123,093  12,309,300      100      100      100
./reads/EXCITED_1_R1.fq  FASTQ   DNA    237,018  23,701,800      100      100      100
./reads/EXCITED_1_R2.fq  FASTQ   DNA    237,018  23,701,800      100      100      100
./reads/EXCITED_2_R1.fq  FASTQ   DNA    158,009  15,800,900      100      100      100
./reads/EXCITED_2_R2.fq  FASTQ   DNA    158,009  15,800,900      100      100      100
./reads/EXCITED_3_R1.fq  FASTQ   DNA    196

**5.2 What is the summary so far?**


 - The experiment captures data from two states: BORED and EXCITED.
 - Three replicates were measured for each state: 1, 2, 3.
 - The sequencing run is paired-end designated by R1 and R2. The read length is 100bp.
 - The measurements are distributed somewhat unevenly over the samples.
 - Twice as much data was collected for sample EXCITED_1 than for BORED_1.
 
 
We have previously identified the genome and transcriptome sizes to be 128,765 and 82,708 respectively. If the coverage were uniform, then the expected coverage of the worst covered sample BORED_1 would be:

112,193 * 100 / 128,765 = 87x
or, we consider the transcriptome alone as target:

112,193 * 100 / 82,708 = 135x
Thus if the expressions were uniform (every gene expressed at the same level), each transcript would be covered at 135x.

Now, we would never expect that gene expression to be uniform for any organism; still the question is how much will the expression deviate from the 135x. As a matter of fact, is the 135x even good as a fairly unreliable approximation? Will most genes have around 135 reads covering each of their coordinate?



**5.3 Generate the root names**



As we pointed out in the Art of Bioinformatics Scripting, every analysis needs to operate on so-called "root" names rather than using "patterns" to match files. Never run any method on patterns like *_R1.fq even if initially it seems to work.

Don't allow the computer to "match" and discover files on its own. Pain and suffering lie that way - all it takes is one extra, or out of order or missing file to potentially render the entire analysis incorrect, often without even a hint that something went amiss. The most serious analysis errors are caused not by applying the wrong methods, but by plugging the wrong data in the wrong place.

**What are the root ids of our problem at hand?**

We see that the reads pairs R1, R2 will need to be handled together. Beyond that, we have the keep the states and replicates separated. Appropriate root ids would be EXCITED_1, EXCITED_2 etc. We can write out these root ids by hand, or generate them with parallel like so:


In [34]:
%%bash

cd ./work/golden/reads
parallel -j 1 echo {1}_{2} ::: BORED EXCITED ::: 1 2 3

BORED_1
BORED_2
BORED_3
EXCITED_1
EXCITED_2
EXCITED_3


We set the -j 1 flag above (number of jobs run at the same time) to limit parallel to executing only one job at a time (in sequential order). Otherwise, some jobs might finish a faster and might produce ids in a different order. The order would not make the ids incorrect, just potentially confusing.

Put these ids into a file:



In [35]:
%%bash

cd ./work/golden/reads
parallel -j 1 echo {1}_{2} ::: BORED EXCITED ::: 1 2 3 > ids


**6 4. Alignment based RNA-Seq**


**6.1 Index your reference genome**


When aligning against a reference genome we first need to prepare (index) that genome so that our software can read it efficiently. This indexing needs to be done only once, after indexing it once the index can be reused when aligning the same organism. We even recommend placing the index outside of the project folder, for clarity we will keep it local:



In [36]:
%%bash

cd ./work/golden

# The reference genome.
IDX=refs/genome.fa

# Build the genome index.
hisat2-build $IDX $IDX

# Index the reference genome with samtools.
samtools faidx $IDX


Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:00:00
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:00
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:00
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Split 1, merged 6; iterating...
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 18392.7 (target: 24140)
Getting block 1 of 7
  Reserving size (24141) for bucket 1
  Calculating Z arrays for bucket 1
  Entering block accumulator loop for bucket 1:
  buck

Settings:
  Output files: "refs/genome.fa.*.ht2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Local offset rate: 3 (one in 8)
  Local fTable chars: 6
  Local sequence length: 57344
  Local sequence overlap between two consecutive indexes: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  refs/genome.fa
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
  Time to read SNPs and splice sites: 00:00:00
Using parameters --bmax 24141 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 24141 --dcv 1024
Constructing suffix-array element gene

Does it matter what you list first in your ids file BORED or EXCITED? Yes.

When we make a pairwise comparison, as a convention, we compare the second category to the first. Our differential expressions will be expressed SECOND/FIRST. While you can, of course, change the order later, it is best if your initial listing already captures the correct order.

In this project, we want to know the prodice the changes of the EXCITED state relative to the BORED state, the fold change is expressed as EXCITED/BORED, thus we list BORED first and EXCITED second.

6.2 Generate the alignments
Since we will generate a BAM alignment for each sample, we'll make a directory to store these alignments and other related files:



In [37]:
%%bash


cd ./work/golden

#Create the BAM folder.
mkdir -p bam



In [39]:
%%bash

cd ./work/golden

IDX=refs/genome.fa

#Align the FASTQ files to the reference genome.
cat ids | parallel "hisat2 -x $IDX -1 ./reads/{}_R1.fq -2 ./reads/{}_R2.fq -S ./bam/{}.sam"


#Sort each SAM into a BAM file.
cat ids | parallel "samtools sort ./bam/{}.sam > ./bam/{}.bam"


#Index each BAM file.
cat ids | parallel  "samtools index ./bam/{}.bam"


#The alignments have all been created, we can see the resulting files with:
ls -l bam



total 781388
-rw-rw-r-- 1 preonath preonath  13015232 Apr  2 00:27 BORED_1.bam
-rw-rw-r-- 1 preonath preonath       352 Apr  2 00:27 BORED_1.bam.bai
-rw-rw-r-- 1 preonath preonath    393565 Mar 28 17:04 BORED_1.bg
-rw-rw-r-- 1 preonath preonath    147516 Mar 28 17:05 BORED_1.bw
-rw-rw-r-- 1 preonath preonath  78719382 Apr  2 00:27 BORED_1.sam
-rw-rw-r-- 1 preonath preonath  16388619 Apr  2 00:27 BORED_2.bam
-rw-rw-r-- 1 preonath preonath       352 Apr  2 00:27 BORED_2.bam.bai
-rw-rw-r-- 1 preonath preonath    424239 Mar 28 17:04 BORED_2.bg
-rw-rw-r-- 1 preonath preonath    155320 Mar 28 17:05 BORED_2.bw
-rw-rw-r-- 1 preonath preonath  96642396 Apr  2 00:27 BORED_2.sam
-rw-rw-r-- 1 preonath preonath  14264773 Apr  2 00:27 BORED_3.bam
-rw-rw-r-- 1 preonath preonath       320 Apr  2 00:27 BORED_3.bam.bai
-rw-rw-r-- 1 preonath preonath    407808 Mar 28 17:04 BORED_3.bg
-rw-rw-r-- 1 preonath preonath    151634 Mar 28 17:05 BORED_3.bw
-rw-rw-r-- 1 preonath preonath  86388077 Apr  2 00:27 BOR

112193 reads; of these:
  112193 (100.00%) were paired; of these:
    319 (0.28%) aligned concordantly 0 times
    111874 (99.72%) aligned concordantly exactly 1 time
    0 (0.00%) aligned concordantly >1 times
    ----
    319 pairs aligned concordantly 0 times; of these:
      316 (99.06%) aligned discordantly 1 time
    ----
    3 pairs aligned 0 times concordantly or discordantly; of these:
      6 mates make up the pairs; of these:
        0 (0.00%) aligned 0 times
        6 (100.00%) aligned exactly 1 time
        0 (0.00%) aligned >1 times
100.00% overall alignment rate
123093 reads; of these:
  123093 (100.00%) were paired; of these:
    371 (0.30%) aligned concordantly 0 times
    122696 (99.68%) aligned concordantly exactly 1 time
    26 (0.02%) aligned concordantly >1 times
    ----
    371 pairs aligned concordantly 0 times; of these:
      363 (97.84%) aligned discordantly 1 time
    ----
    8 pairs aligned 0 times concordantly or discordantly; of these:
      16 mates ma

In [40]:
%%bash


#The alignments have all been created, we can see the resulting files with:

cd ./work/golden/bam

ls -l 

total 781388
-rw-rw-r-- 1 preonath preonath  13015232 Apr  2 00:27 BORED_1.bam
-rw-rw-r-- 1 preonath preonath       352 Apr  2 00:27 BORED_1.bam.bai
-rw-rw-r-- 1 preonath preonath    393565 Mar 28 17:04 BORED_1.bg
-rw-rw-r-- 1 preonath preonath    147516 Mar 28 17:05 BORED_1.bw
-rw-rw-r-- 1 preonath preonath  78719382 Apr  2 00:27 BORED_1.sam
-rw-rw-r-- 1 preonath preonath  16388619 Apr  2 00:27 BORED_2.bam
-rw-rw-r-- 1 preonath preonath       352 Apr  2 00:27 BORED_2.bam.bai
-rw-rw-r-- 1 preonath preonath    424239 Mar 28 17:04 BORED_2.bg
-rw-rw-r-- 1 preonath preonath    155320 Mar 28 17:05 BORED_2.bw
-rw-rw-r-- 1 preonath preonath  96642396 Apr  2 00:27 BORED_2.sam
-rw-rw-r-- 1 preonath preonath  14264773 Apr  2 00:27 BORED_3.bam
-rw-rw-r-- 1 preonath preonath       320 Apr  2 00:27 BORED_3.bam.bai
-rw-rw-r-- 1 preonath preonath    407808 Mar 28 17:04 BORED_3.bg
-rw-rw-r-- 1 preonath preonath    151634 Mar 28 17:05 BORED_3.bw
-rw-rw-r-- 1 preonath preonath  86388077 Apr  2 00:27 BOR

**6.4 Create coverage data**


Now, if you do load the data, you will notice a problem. Even if you only load in a single BAM file, it takes quite a bit of time to have it appear in IGV. Moreover, the same loading process needs to take place after each zooming or panning operation.

It quickly becomes incredibly tedious to use, frankly an embarrassment for the Broad Institute the maintainers of the software. Really? Is this how we are curing cancer? Is this the best tool that hundreds of millions of dollars in public and private funding are able to provide? A genome visualizer that gets bogged down when presented with the Gold Snidget genome?

Of course, there is a workaround. There always is. Sometimes, I feel bioinformatics itself is the journey of working around problems that shouldn't exist in the first place. Since what we primarily care about is the "coverage" rather than individual alignments, we can turn the BAM file into a so-called BigWig file - ah yes, I wish I was joking.

The bigWig format is the brainchild of a celebrated bioinformatician Dr. Jim Kent of the Human Genome fame. Among the many notable accomplishments, Dr. Kent is also the creator of the UCSC genome browser, a tool of high utility. Alas, he is also the "inventor" of ridiculous data formats such as bed or bigWig, formats that are little more than band-aid solutions to much larger, more fundamental, and still unaddressed problems of biological data representation.

Silly or not here we go, bigWig is what must use to work around the limitations of IGV. Oh well, let's make a bigWig then... But first install the bedGraphToBigWig converter if you don't already have it.

In [41]:
%%bash

conda install -y ucsc-bedgraphtobigwig

Solving environment: ...working... done

## Package Plan ##

  environment location: /home/preonath/anaconda3

  added / updated specs:
    - ucsc-bedgraphtobigwig


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-4.10.0               |   py37h89c1867_0         3.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.1 MB

The following packages will be UPDATED:

  conda                                4.9.2-py37h89c1867_0 --> 4.10.0-py37h89c1867_0



conda-4.10.0         | 3.1 MB    | ########## | 100% 
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done


You see you can't just make a bigWig, that would be too easy. First you turn your BAM files into bedGraph then you take each bedGraph and turn it into a bigWig. Sigh... Thankfully we have it all automated. The following steps will be necessary:



In [43]:
%%bash


cd ./work/golden
IDX=refs/genome.fa

#Turn each BAM file into bedGraph coverage. The files will have the .bg extension.
cat ids | parallel "bedtools genomecov -ibam  ./bam/{}.bam -split -bg  > ./bam/{}.bg"

#Convert each bedGraph coverage into bigWig coverage. The files will have the .bw extension.
cat ids | parallel "bedGraphToBigWig ./bam/{}.bg  ${IDX}.fai ./bam/{}.bw"


The resulting bedGraph and bigWig files will have *.bg and *.bw extensions and are placed in the bam directory.

You can drag your bigWig files in the IGV panels, and the coverage information will load up much-much faster. Below I have loaded all samples, re-sized the tracks, colored them by samples. I have also turned on logarithmic and automatic scaling. The resulting browser track is quite informative:



Note that once you make a useful visualization with IGV, you can save the IGV session into a file that can be reloaded later.



**6.5 What can you learn from the visualization?**


Try to answer the following questions from the visualization:

 - What are the minimal gene expression levels that can be detected at all?

 - What are the minimal gene expression levels that could be used to detect a four-fold change?

 - Is the 135x average coverage computed in the previous chapter a reasonable approximation for average gene expression?

 - Look at the gene names. Does the data seem to sort of support the expected gene regulation?

 - What else can you read off the tracks?
6.6 What is our quantification matrix?
The goal of an RNA-Seq method is to produce the quantification matrix. In an alignment-based RNA-Seq, the process requires intersecting the BAM alignment file with the intervals listed in an annotation file.

The intersection process, often called "feature counting", produces a count of how many alignments overlap with each feature of the file. For example, take GENE A as 

pictured below. Assume that each dashed line is an alignment. How many features overlap with GENE A?

-------    -------  -------
   ------  ---------  ---------
     -------  ---------
 ------- --------- ---------
      |<---  GENE A  --->|
      
      
As you can probably note yourself, the answer depends on what the word "overlap" means precisely.

 - Will any amount of overlap suffice? Then the 11 alignments overlap with the feature.

 - Should the alignment be completely inside the feature? Then the count is only 4.

 - Should at least 50% of the alignment cover the feature? Now the count is another value.
 
There is no universally correct way to count; depending on the situation, different strategies might be more appropriate. The default setting that we use will consider any amount of overlap to increase the counts.

**6.7 How to count the features?**


We recommend using the tool featureCounts, many scientists seem to swear by htseq-counts. A typical invocation of featureCounts would be:

In [44]:
%%bash 
cd ./work/golden

#Run the featureCounts program to summarize the number of reads that overlap with features.
featureCounts -p -a ./refs/features.gff -o counts.txt ./bam/BORED_1.bam


        ==========     _____ _    _ ____  _____  ______          _____  
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \ 
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
	  v2.0.1

//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 1 BAM file                                       ||
||                           o BORED_1.bam                                    ||
||                                                                            ||
||             Output file : counts.txt                                       ||
||                 Summary : counts.txt.summary                  

In [45]:
%%bash

cd ./work/golden

# Show the first five lines of the counts.txt file.
cat counts.txt | head -5



# Program:featureCounts v2.0.1; Command:"featureCounts" "-p" "-a" "./refs/features.gff" "-o" "counts.txt" "./bam/BORED_1.bam" 
Geneid	Chr	Start	End	Strand	Length	./bam/BORED_1.bam
AAA-750000-UP-4	Golden	1	1060	+	1060	8103
ABA-187500-UP-4	Golden	1560	2083	+	524	979
ACA-46875-UP-4	Golden	2583	3616	+	1034	505


In [46]:
%%bash

cd ./work/golden


#A nice feature of featureCounts is that we can list multiple BAM files:

featureCounts -p -a ./refs/features.gff -o counts.txt ./bam/BORED_1.bam ./bam/BORED_2.bam ./bam/BORED_3.bam
 
#While in general, we try to avoid pattern matching, in this case, we can save a lot of typing by letting the shell fill in the file names as long as we are quite certain these are in the correct order:

featureCounts -p -a ./refs/features.gff -o counts.txt ./bam/BORED_?.bam ./bam/EXCITED_?.bam
 
#The question mark will expand to all possible combinations, the simplest way to test out what a shell metacharacter will do is to use it in different contexts:




        ==========     _____ _    _ ____  _____  ______          _____  
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \ 
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
	  v2.0.1

//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 3 BAM files                                      ||
||                           o BORED_1.bam                                    ||
||                           o BORED_2.bam                                    ||
||                           o BORED_3.bam                                    ||
||                                                               

In [47]:
%%bash

cd ./work/golden/bam

ls BORED_?.bam

BORED_1.bam
BORED_2.bam
BORED_3.bam


In [48]:
%%bash

cd ./work/golden

#running the command mentioned before:

featureCounts -p -a ./refs/features.gff -o counts.txt ./bam/BORED_?.bam ./bam/EXCITED_?.bam
 
#will create the file text file counts.txt that when opened in Excel will contain:




        ==========     _____ _    _ ____  _____  ______          _____  
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \ 
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
	  v2.0.1

//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 6 BAM files                                      ||
||                           o BORED_1.bam                                    ||
||                           o BORED_2.bam                                    ||
||                           o BORED_3.bam                                    ||
||                           o EXCITED_1.bam                     

**6.8 There is even more to counting**


When we run featureCounts with default parameters, numerous decisions are made on our behalf:

Which types of features are used to count overlaps (default=exons)
How are counted features grouped into the same (default=gene_name attribute)
How much overlap is required to consider an alignment overlapping with a feature (default=1)
Should individual read alignments or read-pairs be counted(default=no, setting the -p will count read pairs)
Depending on the use case, you may need to override one or more of these parameters. See all possible options on the documentation page or by invoking the -h flag:

featureCounts -h
If you do override any parameters, make a note of that in your script, like so:




In [49]:
%%bash

cd ./work/golden

# Count read pairs that overlap on the same strand.
featureCounts -p -S 1 ...

The "-S" option has been depreciated.

Version 2.0.1

Usage: featureCounts [options] -a <annotation_file> -o <output_file> input_file1 [input_file2] ... 

## Mandatory arguments:

  -a <string>         Name of an annotation file. GTF/GFF format by default. See
                      -F option for more format information. Inbuilt annotations
                      (SAF format) is available in 'annotation' directory of the
                      package. Gzipped file is also accepted.

  -o <string>         Name of output file including read counts. A separate file
                      including summary statistics of counting results is also
                      included in the output ('<string>.summary'). Both files
                      are in tab delimited format.

  input_file1 [input_file2] ...   A list of SAM or BAM format files. They can be
                      either name or location sorted. If no files provided,
                      <stdin> input is expected. Location-sorted pa

CalledProcessError: Command 'b'\ncd ./work/golden\n\n# Count read pairs that overlap on the same strand.\nfeatureCounts -p -S 1 ...\n'' returned non-zero exit status 255.

**6.9 What is the next step**


You can now either:

Learn how compute the same quantification using classification based methods.
Continue on to the next step on how to perform differential expression.


**7 5. Classification based RNA-Seq**


A classification based RNA-Seq works via so-called "pseudo-alignments".

Instead of using a genome as a reference, each read is classified as (assigned to) a single transcript. Since some reads can be assigned to multiple transcripts, an internal redistribution algorithm is also run, that will attempt to redistribute the reads across similar regions of transcripts to validate other observed constraints.

**7.1 How does the read redistribution work?**



Imagine your transcriptome has two transcripts that look like so:

AAAAAAAAAAGGGGGGGGG
AAAAAAAAAATTTTTTTTT


Now imagine that:

 - AAAAAAAAAAGGGGGGGGG is expressed with 100 copies but
 - AAAAAAAAAATTTTTTTTT expressed with only 10.
 
 
During sequencing these transcript are broken into small fragments. Thus, after sequencing, suppose we observe:

 - 110 reads that cover of the AAAAAAAAAA region.
 - 100 reads will have GGGGGGGGG in them,
 - 10 reads will cover TTTTTTTTT.
 
 
If we were to simply adding up the reads "naively" as they match the transcripts we would end up with:

 - 200 reads matching AAAAAAAAAAGGGGGGGGG
 - 110 reads matching AAAAAAAAAATTTTTTTTT
 
The AAAAAAA region is double counted, those reads match both transcripts. From these observations and constraints, the classifier will attempt to redistribute the 100 reads that match both transcript in a way that is compatible with differences observed later on 100 and 10. Ideally by the end of the redistribution we do get the correct 1/10 ratio for the counts over the two transcripts.

Of course, in the simplified example above, there is only a single "correct" redistribution that could match both cases. When the transcripts are more complicated, when there are more similar variants that overlap in various ways, it becomes a lot more challenging to figure out what the rationale for a particular classification is.

**7.2 What are the advantages of classification?**


Since the methods operate on the sequences for the transcript, they can be "smarter" when it comes to establishing the accurate abundances. In an alignment-based RNA-Seq, the two steps, alignment, and counting are independent and separate processes; the algorithms are unaware of one another.

In a classification-based method, detecting the "alignment" and the count are the same step. Thus we rightfully expect that the methods be more sensitive.

**7.3 What are the disadvantages of classification?**


The primary limitation of classification-based methods is that we need to present the classifier with matching targets. These targets should cover all cases. If we miss one valid target, then the reads that would match those will get redistributed over all other transcripts - potentially negatively impacting the accuracy of the results.

Missing targets can be easily detected. Whereas if we aligned the reads to the genome, we could visually observe locations that are covered but are yet unannotated.

**7.4 What is the weakest link in classification?**


The difficulty in understanding why a specific transcript is reported as differentially expressed is perhaps the greatest weakness of all classification methods. They behave as black boxes. Say transcript T1, containing one extra-base when compared to transcript T2 comes up differentially expressed. How strong is the evidence for T1 being differentially expressed? The internal process and decision making of the redistribution algorithm is not made clear, nor do tools produce any evidence that will allow us to judge the correctness of the process.

Most classification methods are "faith-based" we blindly trust them (not a good strategy to say the least). In our opinion, all classifier generated results should be investigated with and supported by alignment-based validation.

**7.5 What tools implement pseudo-alignment?**


Originally invented and published by Nicolas L Bray, Harold Pimentel, Páll Melsted and Lior Pachter, Near-optimal probabilistic RNA-seq quantification, Nature Biotechnology 34, 525–527 (2016) the algorithms have two widely used implementations:

 - kallisto
 - salmon
 
 
We will use kallisto in our examples.

**7.6 First step of classification**


Prepare the reference for classification:



In [50]:
%%bash

cd ./work/golden

# Set the shortcuts to reference and index.
REF=./refs/transcripts.fa
IDX=./refs/transcript.idx

# Index the genome with kallisto.
kallisto index -i $IDX $REF


#Make a directory for the results
mkdir -p output

cat ids | parallel kallisto quant -i $IDX -o ./output/{} ./reads/{}_R1.fq ./reads/{}_R2.fq


[build] loading fasta file ./refs/transcripts.fa
[build] k-mer length: 31
[build] warning: clipped off poly-A tail (longer than 10)
        from 91 target sequences
[build] warning: replaced 1 non-ACGUT characters in the input sequence
        with pseudorandom nucleotides
[build] counting k-mers ... done.
[build] building target de Bruijn graph ...  done 
[build] creating equivalence classes ...  done
[build] target de Bruijn graph has 92 contigs and contains 77828 k-mers 


[quant] fragment length distribution will be estimated from the data
[index] k-mer length: 31
[index] number of targets: 92
[index] number of k-mers: 77,828
[index] number of equivalence classes: 92
[quant] running in paired-end mode
[quant] will process pair 1: ./reads/BORED_1_R1.fq
                             ./reads/BORED_1_R2.fq
[quant] finding pseudoalignments for the reads ... done
[quant] processed 112,193 reads, 112,193 reads pseudoaligned
[quant] estimated average fragment length: 172.778
[   em] quanti

**7.7 Running the classification**

Running the kallisto classification is surprisingly straightforward task:



In [51]:
%%bash

cd ./work/golden

#Make a directory for the results
mkdir -p output



In [53]:
%%bash
cd ./work/golden
IDX=./refs/transcript.idx

cat ids | parallel kallisto quant -i $IDX -o ./output/{} ./reads/{}_R1.fq ./reads/{}_R2.fq



[quant] fragment length distribution will be estimated from the data
[index] k-mer length: 31
[index] number of targets: 92
[index] number of k-mers: 77,828
[index] number of equivalence classes: 92
[quant] running in paired-end mode
[quant] will process pair 1: ./reads/BORED_1_R1.fq
                             ./reads/BORED_1_R2.fq
[quant] finding pseudoalignments for the reads ... done
[quant] processed 112,193 reads, 112,193 reads pseudoaligned
[quant] estimated average fragment length: 172.778
[   em] quantifying the abundances ... done
[   em] the Expectation-Maximization algorithm ran for 52 rounds


[quant] fragment length distribution will be estimated from the data
[index] k-mer length: 31
[index] number of targets: 92
[index] number of k-mers: 77,828
[index] number of equivalence classes: 92
[quant] running in paired-end mode
[quant] will process pair 1: ./reads/BORED_3_R1.fq
                             ./reads/BORED_3_R2.fq
[quant] finding pseudoalignments for the reads .

In [54]:
%%bash

cd ./work/golden/output
find . -name 'abundance.tsv'


./BORED_3/abundance.tsv
./EXCITED_3/abundance.tsv
./EXCITED_1/abundance.tsv
./BORED_2/abundance.tsv
./EXCITED_2/abundance.tsv
./BORED_1/abundance.tsv


If you really want to know who to blame this oversight for, it was a tool called TopHat first introduced this incredibly counterproductive behavior ... every alignment file produced by Tophat is called accepted.bam. TopHat as an aligner has been long forgotten, superseded by more accurate and better-performing algorithms, but the bad habits introduced with TopHat endure far longer.

Encoding data identity information in the file path is a terrible idea, one we dearly hope won't catch on even more. Above we now have six files, all called 'abundances.tsv' in different folders. Each abundance file has the following structure:

In [56]:
%%bash

cd ./work/golden
cat ./output/BORED_1/abundance.tsv | head -5

target_id	length	eff_length	est_counts	tpm
AAA-750000-UP-4	1059	887.222	8103	51802.2
ABA-187500-UP-4	523	351.222	979	15810.2
ACA-46875-UP-4	1033	861.222	505	3325.92
ADA-23438-UP-4	1022	850.222	206	1374.26


You can read more about effective lengths and tpm on the RNA-Seq terminology page. To make all our statistical methods work the same way, we want to turn the files above into a single counts file similar to what we obtain when we perform an Alignment based RNA-Seq approach.

**7.8 Combine your counts**


Turning the six files into a single count file, where each column is named by sample, is surprisingly challenging to do at the command line.

Now we could always copy-paste by hand, but that act breaks automation. Copy-pasting is an error-prone and tedious process; it takes away our ability to redo an analysis quickly.

It goes to show how a single ill an advised approach, seemingly nothing to do with the algorithm itself, hurts science in general. It compounds problems and makes analyses unnecessarily tricky. When the kallisto developers chose their myopic approach, they have also taken away the ability of people to analyze the data without making use of yet another piece of code.

We'll keep our fingers crossed that the situation will get resolved, but in the meantime, we wrote a Python script to combine all counts into one, a script that we believe is generic enough for all use-cases:

In [58]:
%%bash

cd ./work/golden

#Download the custom script to combine kallisto outputs.

curl http://data.biostarhandbook.com/books/rnaseq/code/combine.py > ~/bin/combine

#Make the script executable.
chmod +x ~/bin/combine

#We can now combine the abundances into a single count matrix with:

 cat ids | combine output > counts.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2369  100  2369    0     0   2864      0 --:--:-- --:--:-- --:--:--  2861


The script inserted three empty columns called X1, X2, X3 to match the structure of the file produced by featureCounts thus the same statistical analysis can be used later on.

You can now continue with The differential expression




8.1 Prepare the environment
If you haven't already done so install the statistical packages into your environment with the command:

**Activate the bioinformatics environment.**


 - conda activate base

In [59]:
%%bash
cd ./work/golden

# Install the statistical packages.
URL=http://data.biostarhandbook.com/books/rnaseq/code/install-conda.txt
curl $URL | xargs conda install -y

Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/preonath/anaconda3

  added / updated specs:
    - bioconductor-deseq
    - bioconductor-deseq2
    - bioconductor-edger
    - r-gplots


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bioconductor-annotate-1.50.0|                0         1.8 MB  bioconda
    bioconductor-annotationdbi-1.32.3|                0         4.4 MB  bioconda
    bioconductor-biobase-2.32.0|                0         1.6 MB  bioconda
    bioconductor-biocgenerics-0.18.0|         r3.2.2_0         502 KB  bioconda
    bioconductor-biocparallel-1.4.3|         r3.2.2_0         638 KB  bioconda
    bioconducto

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    67  100    67    0     0     88      0 --:--:-- --:--:-- --:--:--    88


For this book, we have written scripts that make use of several different R based methods. We have devoted a substantial effort to standardizing both the usage and the results produced by these different methods.

To make these scripts universally accessible and simpler to use, we recommend saving them into the ~/bin folder and making them executable. You can achieve that from the command line with:

In [79]:
%%bash

cd ./work/golden

# Download the scripts into the ~/bin folder.
curl http://data.biostarhandbook.com/books/rnaseq/code/deseq1.r > ~/bin/deseq1.r
curl http://data.biostarhandbook.com/books/rnaseq/code/deseq2.r > ~/bin/deseq2.r
curl http://data.biostarhandbook.com/books/rnaseq/code/edger.r > ~/bin/edger.r
curl http://data.biostarhandbook.com/books/rnaseq/code/heatmap.r > ~/bin/heatmap.r


# Make them executable
chmod +x ~/bin/*.r
pwd
cd ~/bin
deseq1.r
pwd



/home/preonath/Biostar/RNA_Seq/work/golden
/home/preonath/bin


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3466  100  3466    0     0   5180      0 --:--:-- --:--:-- --:--:--  5173
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4171  100  4171    0     0   5916      0 --:--:-- --:--:-- --:--:--  5907
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4148  100  4148    0     0   5908      0 --:--:-- --:--:-- --:--:--  5908
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1874  100  1874    0     0   2661      0 --:--:-- --:--:-- --:--:--  2658
/home/preonath/anaconda3/lib/R/bin/exec/R: error whi

In [80]:
%%bash

cd ~/bin
pwd

deseq1.r
pwd

/home/preonath/bin
/home/preonath/bin


/home/preonath/anaconda3/lib/R/bin/exec/R: error while loading shared libraries: libreadline.so.6: cannot open shared object file: No such file or directory
